### ISSUE TITLE
Stability & Maintenance

### ISSUE THREAD

**Issue by skipjack** [dsheiko/puppetry#84]
@dsheiko first off, thanks so much for your work on this. I've now had a few days to play around with puppetry and it's been the most enjoyable end-to-end testing experience I've ever had (I'm coming from a ui engineering, including some jest + puppeteer, background). I'm evaluating the tool alongside a few others and wanted to first ask some questions, are you...

- Still planning to maintain this and keep it up to date with Jest/Puppeteer?
- ☝️ I see the last release was 8/10/2020 and it looks like work [slowed in early 2020][3].
- Accepting PRs? Good chance I'd submit some fixes for things if we do go with Puppetry.
- Still working on the v4/next projects?

And second, note some things I found while experimenting...

- Exported Jest/Puppeteer project uses `puppeteer@3` while that project has now jumped to v9.
- Firefox fails for file uploads, though this may be [unrelated to this project][1].
- Edge [now works][2] with `puppeteer` and would be cool to include in the browser options (I was able to hack it in).
- It would be cool to store hidden template variables outside the `.puppetryrc`, this would allow them to be `.gitignore`d.

Happy to break out any of the above noted items into separate issues if you'd like.

[1]: https://github.com/puppeteer/puppeteer/issues/3762
[2]: https://docs.microsoft.com/en-us/microsoft-edge/puppeteer/
[3]: https://github.com/dsheiko/puppetry/graphs/contributors

**Comment by dsheiko** [comment]
The work froze more than a year ago when I was working on v4. I had back then some personal problems. I still fancy the idea to revive the project. Actually now I could spare some time, but what holds me back - the understanding how much  work it would require to test/QA the entire project for v4. Another issue is: I don't know how many people really use Puppetry, is it really in demand. My colleges for example didn't find it that exciting and go with Cypress instead. I somehow didn't manage to engage any devs to the project - what's a failure 

Yet, you brought up pretty clear and fair points in here. I'm going to solve them in a new version in branch 3

**Comment by skipjack** [comment]
Thanks for the detailed response!

> The work froze more than a year ago when I was working on v4. I had back then some personal problems.

Yeah I know how it goes, finding time for open source between other things can be tough.

> I still fancy the idea to revive the project. Actually now I could spare some time, but what holds me back - the understanding how much work it would require to test/QA the entire project for v4.

Good to hear that you're still interested in keeping it alive. Would love to help with some of it. Can you elaborate on this part: "the understanding how much work it would require to test/QA the entire project for v4"? I generally get what you mean; QA-ing the QA tool 😁 . I have some ideas but it would be good to know the specific pain points you're referring to.

> Another issue is: I don't know how many people really use Puppetry, is it really in demand. My colleges for example didn't find it that exciting and go with Cypress instead. I somehow didn't manage to engage any devs to the project - what's a failure

Hey almost 700 stars isn't failure, quite the opposite I would say. Another engineer referred us to this project and given the amount of stars clearly some folks are using it. Even despite the issues I mention above, this tool already covers 80% or more of our use cases. The biggest value imho is that _it's approachable for non-engineers_, though you still provide the ability to go low-level when necessary (e.g. exporting). Don't be dissuaded by the naysayers, maybe you were just pitching it to the wrong audience 😁 

> Yet, you brought up pretty clear and fair points in here. I'm going to solve them in a new version in branch 3.

Again, let me know how I can help (even if it's just brainstorming). You're on to something with this project and visual builders like this one that are approachable to folks who aren't full-on engineers are where the industry is heading (e.g. Testim).

**Comment by dsheiko** [comment]
Any help would be pretty much appreciated. As for the focus (backlog):
Besides the update to match new version Puppeteer I keep in mind for later the 4.x branch:
Here what was in progress
1) UI update ![puppetry](https://user-images.githubusercontent.com/1393963/117027266-d4fcb780-acfc-11eb-973d-e948c111607e.PNG)
2) Support for multiple window
3) Batch testing with Spider https://www.facebook.com/puppetry.testing/posts/133940964699171

Tell me pls what would you see as higher priority

**Comment by skipjack** [comment]
Ok so...

> UI update

Looks good! I like the delineation between navigation bars (side panel for projects, top bar for global settings). Some questions though...

- What are "Windows"? Sounds cool 😁 
- If a "project" is one to one with a git repo, it feel likes the targets and snippets (even if shared between suites within the project) should be scoped to that project? Maybe that's how it already works but having it as a separate tab in the side panel makes me think differently.

I have some other UI niceties to suggest but I can do this separately (e.g. on a mac, cmd quit doesn't work as expected and the app doesn't have a normal close/minimize/maximize menu bar).

> Support for multiple window

I'm guessing this is that "Windows" tab? Can you elaborate?

> Batch testing with Spider

I'm not on Facebook so I can't see that post. Can you elaborate on this one too? I see a few different "spider" tools, not sure which on you're referring to.

**Comment by skipjack** [comment]
> Tell me pls what would you see as higher priority

Here's what I'd recommend now that you've confirmed you're still planning to work on this...

- [ ] I'll close this issue and create some more granular ones (e.g. one for bumping the puppeteer version).
- [ ] Consolidate the [two projects][1] into one so it's easier to track.
- [ ] Close/delete whatever issues/notes are not applicable any more.
- [ ] Convert any project notes that are still applicable into issues for better visibility.
- [ ] Prioritize things (to answer your question I'd say the puppeteer bump should be highest) and then start cutting pre-releases (`v4.0.0-next.0`, `v4.0.0-next.1`, etc) as we knock things down.

At the end of the day this is totally up to you though, just tell me what you agree with and I'd be happy to help. I already started poking around with a puppeteer bump (straight to v9) and for the most part it worked perfectly, can share details when we create an isolated issue for it.

[1]: https://github.com/dsheiko/puppetry/projects

**Comment by dsheiko** [comment]
https://github.com/dsheiko/puppetry/releases/tag/v3.2.5

**Comment by dsheiko** [comment]
I'm thinking about to flatten structure group/test/actions to test/actions. It is supposed to improve overall performance, UI response and probably be more convenient to users. Puppetry will convert older suits into V4 structure automatically (probably changing test names into group name: test name). What do you think?

**Comment by skipjack** [comment]
They'd still run consecutively? It's only entire suites that run in parallel, correct?

**Comment by dsheiko** [comment]
"They'd still run consecutively? It's only entire suites that run in parallel, correct?" - correct
"Consolidate the two projects into one so it's easier to track." - Puppetry 4 - is one I'm working on, Puppetry Next - backlog including what I didn't include in 4th
Window support - currently all the tests run in a single window. If it opens a new window, the tests cannot follow the opened one. I want to add to the target window where to run the test.
Spider - when set up, the crawler indexes pages of the target site and we can set "run this test on all the pages like /foo/**/*"

![p1](https://user-images.githubusercontent.com/1393963/118112954-f8eb8780-b3e5-11eb-8443-4b1a4f35e0be.png)
![p2](https://user-images.githubusercontent.com/1393963/118112965-fbe67800-b3e5-11eb-8376-8694a72fb71c.png)

**Comment by skipjack** [comment]
> "They'd still run consecutively? It's only entire suites that run in parallel, correct?" - correct

Awesome, then yea I agree flattening makes sense.

> "Consolidate the two projects into one so it's easier to track." - Puppetry 4 - is one I'm working on, Puppetry Next - backlog including what I didn't include in 4th

Ok, I'd recommend dropping "Puppeteer Next" as it seems there's some duplication between the two. For example, windows and crawling are mentioned in both. Anything that's not planned for Puppetry 4 but that you still want to track can just be an issue. Again, your call but I think this would help us to focus on v4 and give other contributors who look at the project a clear picture of what's happening.

> Window support - currently all the tests run in a single window. If it opens a new window, the tests cannot follow the opened one. I want to add to the target window where to run the test.

Makes sense, I like this idea too.

> Spider - when set up, the crawler indexes pages of the target site and we can set "run this test on all the pages like /foo/**/*"

Love this, I hadn't even thought of running tests like this but it makes a lot of sense. Would it handle SPA-like sites? Is there a tool you'd be using under the hood? If so, can you share the link?

**Comment by skipjack** [comment]
I see the `puppeteer` to v9 bump as well 🎉

**Comment by skipjack** [comment]
Created #85, #86, and #87. I would say #86 is the most important, please let me know if you're open to these ideas and how I can help.
